File: artifacts/api-server/src/lib/logoProcessor.ts
import sharp from "sharp";

export interface LogoVariants {
  original: string;
  black: string;
  white: string;
  grayscale: string;
}

/**
 * Generate logo variants from a base64 data URL or Buffer.
 * Returns { black, white, grayscale } as PNG Buffers.
 */
export async function generateLogoVariants(
  inputBuffer: Buffer
): Promise<{ black: Buffer; white: Buffer; grayscale: Buffer }> {
  const { data, info } = await sharp(inputBuffer)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width, height } = info;
  const total = width * height * 4;

  const blackPx = Buffer.alloc(total);
  const whitePx = Buffer.alloc(total);
  const grayPx = Buffer.alloc(total);

  for (let i = 0; i < total; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    const a = data[i + 3];
    const gray = Math.round(0.299 * r + 0.587 * g + 0.114 * b);

    blackPx[i] = 0;
    blackPx[i + 1] = 0;
    blackPx[i + 2] = 0;
    blackPx[i + 3] = a;

    whitePx[i] = 255;
    whitePx[i + 1] = 255;
    whitePx[i + 2] = 255;
    whitePx[i + 3] = a;

    grayPx[i] = gray;
    grayPx[i + 1] = gray;
    grayPx[i + 2] = gray;
    grayPx[i + 3] = a;
  }

  const rawOpts = { raw: { width, height, channels: 4 as const } };

  const [black, white, grayscale] = await Promise.all([
    sharp(blackPx, rawOpts).png().toBuffer(),
    sharp(whitePx, rawOpts).png().toBuffer(),
    sharp(grayPx, rawOpts).png().toBuffer(),
  ]);

  return { black, white, grayscale };
}

/**
 * Convert a base64 data URL to a Buffer.
 */
export function dataUrlToBuffer(dataUrl: string): Buffer {
  const base64 = dataUrl.replace(/^data:[^;]+;base64,/, "");
  return Buffer.from(base64, "base64");
}

/**
 * Check if an image buffer has any transparent pixels (alpha < 128).
 */
export async function hasTransparency(buffer: Buffer): Promise<boolean> {
  const { data, info } = await sharp(buffer)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });
  for (let i = 3; i < data.length; i += 4) {
    if (data[i] < 128) return true;
  }
  return false;
}

/**
 * Extract dominant colors from a logo buffer, ignoring transparent and near-white/black pixels.
 */
export async function extractLogoColors(buffer: Buffer, numColors = 6): Promise<string[]> {
  const { data, info } = await sharp(buffer)
    .ensureAlpha()
    .resize(80, 80, { fit: "inside" })
    .raw()
    .toBuffer({ resolveWithObject: true });

  const buckets: Record<string, { r: number; g: number; b: number; count: number }> = {};

  for (let i = 0; i < data.length; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    const a = data[i + 3];

    if (a < 128) continue;
    const lum = (0.299 * r + 0.587 * g + 0.114 * b) / 255;
    if (lum < 0.05 || lum > 0.97) continue;

    const rB = Math.round(r / 24) * 24;
    const gB = Math.round(g / 24) * 24;
    const bB = Math.round(b / 24) * 24;
    const key = `${rB},${gB},${bB}`;
    if (!buckets[key]) buckets[key] = { r: rB, g: gB, b: bB, count: 0 };
    buckets[key].count++;
  }

  const sorted = Object.values(buckets).sort((a, b) => b.count - a.count);
  const selected: string[] = [];

  for (const bucket of sorted) {
    if (selected.length >= numColors) break;
    const hex = toHex(bucket.r, bucket.g, bucket.b);
    const isDup = selected.some((c) => colorDist(c, hex) < 50);
    if (!isDup) selected.push(hex);
  }

  return selected;
}

function toHex(r: number, g: number, b: number): string {
  return `#${[r, g, b].map((v) => Math.min(255, v).toString(16).padStart(2, "0")).join("")}`;
}

function colorDist(h1: string, h2: string): number {
  const parse = (h: string) => {
    const n = parseInt(h.replace("#", ""), 16);
    return [(n >> 16) & 255, (n >> 8) & 255, n & 255];
  };
  const [r1, g1, b1] = parse(h1);
  const [r2, g2, b2] = parse(h2);
  return Math.sqrt((r1 - r2) ** 2 + (g1 - g2) ** 2 + (b1 - b2) ** 2);
}

File: artifacts/api-server-python/app/services/logo_processor.py
"""
Logo processing — generate black/white/grayscale variants and extract dominant colors.
Uses Pillow (already installed). No external dependencies needed.
"""
import base64
import io
import math
from typing import Optional

from PIL import Image


def data_url_to_image(data_url: str) -> Image.Image:
    """Convert a base64 data URL to a Pillow Image."""
    base64_data = data_url.split(",", 1)[-1]
    image_bytes = base64.b64decode(base64_data)
    return Image.open(io.BytesIO(image_bytes)).convert("RGBA")


def image_to_png_bytes(img: Image.Image) -> bytes:
    """Convert a Pillow Image to PNG bytes."""
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    return buf.getvalue()


def generate_logo_variants(img: Image.Image) -> dict[str, bytes]:
    """
    Generate black, white, and grayscale variants of a logo image.
    Preserves alpha channel (transparency).
    Returns {black, white, grayscale} as PNG bytes.
    """
    img = img.convert("RGBA")
    r_data, g_data, b_data, a_data = img.split()

    # Black variant — keep alpha, set RGB to 0
    black = Image.new("RGBA", img.size, (0, 0, 0, 0))
    black.putalpha(a_data)

    # White variant — keep alpha, set RGB to 255
    white = Image.new("RGBA", img.size, (255, 255, 255, 0))
    white.putalpha(a_data)

    # Grayscale variant — luminosity-weighted gray, keep alpha
    gray_img = img.convert("LA").convert("RGBA")

    return {
        "black": image_to_png_bytes(black),
        "white": image_to_png_bytes(white),
        "grayscale": image_to_png_bytes(gray_img),
    }


def extract_logo_colors(img: Image.Image, num_colors: int = 6) -> list[str]:
    """
    Extract dominant colors from a logo, ignoring transparent and near-white/black pixels.
    Returns a list of hex color strings.
    """
    img = img.convert("RGBA").resize((80, 80), Image.LANCZOS)
    pixels = list(img.getdata())

    buckets: dict[str, dict] = {}
    for r, g, b, a in pixels:
        if a < 128:
            continue
        lum = (0.299 * r + 0.587 * g + 0.114 * b) / 255
        if lum < 0.05 or lum > 0.97:
            continue

        # Quantize to buckets of 24
        rb = round(r / 24) * 24
        gb = round(g / 24) * 24
        bb = round(b / 24) * 24
        key = f"{rb},{gb},{bb}"
        if key not in buckets:
            buckets[key] = {"r": rb, "g": gb, "b": bb, "count": 0}
        buckets[key]["count"] += 1

    sorted_buckets = sorted(buckets.values(), key=lambda x: x["count"], reverse=True)
    selected: list[str] = []

    for bucket in sorted_buckets:
        if len(selected) >= num_colors:
            break
        hex_color = _to_hex(bucket["r"], bucket["g"], bucket["b"])
        if not any(_color_dist(hex_color, c) < 50 for c in selected):
            selected.append(hex_color)

    return selected


def _to_hex(r: int, g: int, b: int) -> str:
    return f"#{min(r, 255):02x}{min(g, 255):02x}{min(b, 255):02x}"


def _color_dist(h1: str, h2: str) -> float:
    def parse(h: str):
        n = int(h.lstrip("#"), 16)
        return (n >> 16) & 255, (n >> 8) & 255, n & 255

    r1, g1, b1 = parse(h1)
    r2, g2, b2 = parse(h2)
    return math.sqrt((r1 - r2) ** 2 + (g1 - g2) ** 2 + (b1 - b2) ** 2)
